File: framework/languages/dotnet/e2e/RegistryMessaging/Client/Scenarios/RmA4SameRidFailoverScenario.cs
using Zlink.HttpClient;
using RegistryMessaging.Shared;
using RegistryMessaging.Client.Support;

namespace RegistryMessaging.Client.Scenarios;

// RM-A4 verifies that discovery keeps using the same logical provider rid
// after the backing provider process is replaced with a new endpoint.
internal static class RmA4SameRidFailoverScenario
{
    public static async Task RunAsync(ClientOptions options)
    {
        await using var cluster = await DynamicClusterLauncher.StartAsync(options, "rm-a4");
        var providerV1 = await cluster.StartProviderAsync("api-a-v1", "api-a");
        using var providerV1Client = ZLinkHttpClient.Create(providerV1.HttpUrl)
            .Json()
            .Timeout(TimeSpan.FromMinutes(5))
            .Build();

        var first = (await providerV1Client.Post("/profile/request")
            .Body(new ProfileRequest("rm-a4-v1"))
            .SubmitAsync<ProfileReply>()).Body;
        ScenarioAssert.That(
            first.ProviderRid == "api-a",
            "RM-A4 initial request should reach api-a.");

        await WaitForEvidenceAsync(providerV1Client, "value=rm-a4-v1");

        await cluster.StopAsync(providerV1);

        var providerV2 = await cluster.StartProviderAsync("api-a-v2", "api-a");
        using var providerV2Client = ZLinkHttpClient.Create(providerV2.HttpUrl)
            .Json()
            .Timeout(TimeSpan.FromMinutes(5))
            .Build();

        var beforeV1 = await ReadEvidenceIgnoringStoppedAsync(providerV1Client);
        var beforeV2 = await ReadEvidenceAsync(providerV2Client);
        var marker = $"rm-a4-{Guid.NewGuid():N}";
        for (var i = 0; i < 20; i++)
        {
            var reply = (await providerV2Client.Post("/profile/request")
                .Body(new ProfileRequest($"{marker}-{i}"))
                .SubmitAsync<ProfileReply>()).Body;
            ScenarioAssert.That(
                reply.ProviderRid == "api-a",
                "RM-A4 replacement request should reach api-a.");
        }

        var afterV2 = await WaitForEvidenceAsync(providerV2Client, $"{marker}-19");
        var afterV1 = await ReadEvidenceIgnoringStoppedAsync(providerV1Client);
        var v1Count = ScenarioAssert.CountNewEvidence(
            afterV1,
            beforeV1,
            "profile-request|rid=api-a",
            marker);
        var v2Count = ScenarioAssert.CountNewEvidence(
            afterV2,
            beforeV2,
            "profile-request|rid=api-a",
            marker);
        ScenarioAssert.That(v1Count == 0 && v2Count == 20, "RM-A4 replacement provider evidence did not match.");

        Console.WriteLine("scenario RM-A4 passed");
    }

    static async Task<string[]> ReadEvidenceAsync(ZLinkHttpClient client) =>
        (await client.Get("/evidence").SubmitAsync<string[]>()).Body;

    static async Task<string[]> ReadEvidenceIgnoringStoppedAsync(ZLinkHttpClient client)
    {
        try
        {
            return await ReadEvidenceAsync(client);
        }
        catch
        {
            return [];
        }
    }

    static async Task<string[]> WaitForEvidenceAsync(ZLinkHttpClient client, string contains) =>
        (await client.Post("/evidence/wait")
            .Body(new EvidenceWaitRequest(contains))
            .SubmitAsync<string[]>()).Body;
}

File: framework/languages/node/e2e/RegistryMessaging/Client/Scenarios/rm-a4-same-rid-failover-scenario.ts
import type { ClientOptions } from '../Support/client-options';
import { DynamicClusterLauncher } from '../Support/dynamic-cluster-launcher';
import { getJson, postJson } from '../Support/http-client';
import { countNewEvidence, ensure, uniqueMarker } from '../Support/scenario-assert';
import type { ProfileReply } from '../../Shared/messages';

export async function runRmA4(options: ClientOptions): Promise<void> {
  const cluster = await DynamicClusterLauncher.start(options, 'rm-a4');
  try {
    const providerV1 = await cluster.startProvider('api-a-v1', 'api-a');
    const first = await postJson<ProfileReply>(providerV1.httpUrl, '/profile/request', { value: 'rm-a4-v1' });
    ensure(first.providerRid === 'api-a', 'RM-A4 initial request should reach api-a.');
    await postJson<string[]>(providerV1.httpUrl, '/evidence/wait', { contains: 'value=rm-a4-v1' });

    await cluster.stop(providerV1);
    const providerV2 = await cluster.startProvider('api-a-v2', 'api-a');
    const beforeV1 = await readEvidenceIgnoringStopped(providerV1.httpUrl);
    const beforeV2 = await getJson<string[]>(providerV2.httpUrl, '/evidence');
    const marker = uniqueMarker('rm-a4');
    for (let i = 0; i < 20; i += 1) {
      const reply = await postJson<ProfileReply>(providerV2.httpUrl, '/profile/request', { value: `${marker}-${i}` });
      ensure(reply.providerRid === 'api-a', 'RM-A4 replacement request should reach api-a.');
    }
    const afterV2 = await postJson<string[]>(providerV2.httpUrl, '/evidence/wait', { contains: `${marker}-19` });
    const afterV1 = await readEvidenceIgnoringStopped(providerV1.httpUrl);
    const v1Count = countNewEvidence(afterV1, beforeV1, 'profile-request|rid=api-a', marker);
    const v2Count = countNewEvidence(afterV2, beforeV2, 'profile-request|rid=api-a', marker);
    ensure(v1Count === 0 && v2Count === 20, 'RM-A4 replacement provider evidence did not match.');
    console.log('scenario RM-A4 passed');
  } finally {
    await cluster.close();
  }
}

async function readEvidenceIgnoringStopped(url: string): Promise<string[]> {
  try {
    return await getJson<string[]>(url, '/evidence');
  } catch {
    return [];
  }
}
